### ISSUE TITLE
Allow R `NULL` with Option

### ISSUE THREAD

**Issue by dfalbel** [extendr/extendr#138]
Currently when passing a NULL to an `Option<T>`, for example: https://github.com/extendr/extendr/blob/master/extendr-api/src/robj/from_robj.rs#L172-L182

I get an error:

```
expected an integer scalar
```

Should we allow NULL's here and return `None` in this case as we do with `NA's`? IMHO, it's more common to use `NULL` than `NA` when you want your argument to be optional.

More generally, I wonder if it would make sense to have an `ROption` enum, that would distinct NULL's from NA's, eg:

```
enum ROption<T> {
    None,
    Some(T),
    Na
}
```

**Comment by clauswilke** [comment]
In my opinion `NULL` and `NA` live in different worlds and thus don't belong into one option. `NA` indicates absence of a value in a vector. `NULL` indicates absence of the entire object.

Not sure how to best map this onto Rust data types, though. For scalar inputs, mapping `NULL` onto `None` seems fine. For vector inputs, I'm not sure. `NULL` would correspond to a vector of length 0, whereas `NA` would correspond to a vector of length 1 containing no value?

**Comment by andy-thomason** [comment]
We could have a Nullable type that can wrap anything. We could then do a
blanket wrap.

For optional parameters, I'm looking at a way of putting default values
into r wrappers. You could choose NA or null. There is also a special
missing arg value we could use.

**Comment by dfalbel** [comment]
@clauswilke ! Agree that they are different concepts and shouldn't live in the same place.

Considering `Option` for `NULL` we could have something similar to what we have for logical values [here](https://github.com/extendr/extendr/blob/master/extendr-api/src/logical.rs) for all the other types. 

Thus what we would have is something like:

```
Option<bool> # accepts `NULL` or logical values (not NA)
Option<RBool> # accepts `NULL` or logical values including NA_LOGICAL

Vec<bool> # vector of logical (don't accept NA's)
Vec<RBool> # vector of logical (accepting NA's)
Option<Vec<bool>> accepts NULL or a vector of with no NA's
Option<Vec<RBool>> aceepts NULL or a vector of logical (possibly including NA_LOGICAL)
Option<Vec<Option<RBool>>> # errors, or probably not implemented in extendr-api.
```

Would that make sense?

**Comment by clauswilke** [comment]
We wanted to avoid introducing R specific types (`RBool`) as much as possible. The Rust equivalent to "one scalar value of type T or a missing value" is `Option<T>`, so that should be the innermost type, where `Some<T>()` corresponds to an existing value and `None()` corresponds to `NA`.

`NULL` can be handled by additional wrappers around this option type (or vectors thereof), as @andy-thomason suggested.

I also feel it's important to not lose sight of the overarching goal: Making the Rust side as simple, convenient, and idiomatic as possible. If this means some R functions require `NA` instead of `NULL` as input, so be it. One could always write a wrapper on the R side that translates `NULL` to `NA`.

**Comment by clauswilke** [comment]
To be clear: Users of `#[extendr]` could then pick what kind of inputs on the R side exactly they're willing to support:

```
// Must provide exactly one non-missing integer
#[extendr]
fn A(x: int32) {}

// Must provide exactly one integer value or NA
#[extendr]
fn B(x: Option<int32>) {}

// Must provide exactly one integer value or NA, or alternatively NULL for no input
#[extendr]
fn C(x: NullWrap<Option<int32>>) {}
```

**Comment by andy-thomason** [comment]
We already have a Bool type which is widely supported.

For example

```
TRUE  is Bool(1)
FALSE is Bool(0)
NA_BOOLEAN is Bool(0x8000000)
```

There is also a special string na_str() which has the value "NA" but a
special address.

Both Bool and &str work with the IsNA trait.

You will have to use the latest version of the code as the current crate is
out of date.

Andy.

**Comment by andy-thomason** [comment]
If possible, get used to using slices instead of vectors as vectors require
a memory allocation and copy which
is a very bad thing,

**Comment by andy-thomason** [comment]
https://extendr.github.io/extendr/extendr_api/struct.Bool.html

**Comment by dfalbel** [comment]
Ok! That sounds reasonable to me! 

In my mental model of Rust `Option<>` maps more to `NULL` than `NA`, so I would rather have `Option<MaybeNA<i32>>`... but I can totally live with: `NullWrap<Option<i32>>`.

Thanks I'll have a look at slices!

Would you be interested in a PR implementing a NullWrap enum?

**Comment by clauswilke** [comment]
> Would you be interested in a PR implementing a NullWrap enum?

Yes. Not wedded to the name, though. Maybe `MaybeNull`?

**Comment by Ilia-Kosenkov** [comment]
@andy-thomason, could you please clarify a little bit about `Bool`?
From @clauswilke 's example, R vector types can be mapped to primitive types in Rust (say, a scalar) if we disallow `NA`, or to `Option<primitive_type>` if `NA` is allowed, which is perfectly reasonable. 
In such a case, isn't it better to have `Option<bool>` to represent R's `logical`? Is there a performance consideration or some other reason?

**Comment by andy-thomason** [comment]
The Bool exists primarily for performance as it is 32 bits like the logical type.

We do support bool and Option<bool> but they require a copy to work and so is likely to be significantly
slower (10x-1000x).

The same is true of Option<i32> and Option<f64>.

For scalar inputs to functions it would make very little difference, but for large (>1GB) vectors,  it would be significant.

**Comment by Ilia-Kosenkov** [comment]
@andy-thomason,
I suggest to go the `cpp11` way, i.e. to have dedicated `Logical`, `Integer`, `Real`, etc types.
That way we can provide high-performance scalar and slice support, 
```Rust
fn rust_efficient(data : &[Real], flag : Logical) -> Real { todo!() }
```
And using `FromRobj` we can extend conversions to support native types, `Option<>`, `Vec<>` and so on.
I have prototyped it (at least for logical) in my fork and it looks doable and, most importantly, very convenient for the consumer.

I would like to draw your attention to #137: we need a convention for error reporting, right now we use `Result<Self, &str>` in `FromRobj`, and fill in error messages by hand, duplicating (with potential mistakes) text in many places.

I experimented with extending `Error` enum in `extendr_api::error`, adding validation-specific members, and improving reporting of error messages. 
For instance, instead of text, `FromRobj` can emit `Error::ScalarNA` and `Error::ScalarLen(usize)`, which then can be easily converted into meaningful error messages, `{tidyverse}`-style. 
See something like this
https://github.com/Ilia-Kosenkov/extendr/blob/480dfa26d3a8091213474c2eb3dd871b8396b959/extendr-api/src/robj/from_robj.rs#L76
for reference (do not mind my poor Rust implementation).

And we need a convention on `NULL` handling, right now I do not even know what happens if `NULL` gets passed where solid vector type is expected. Something like (not sure how to write `FromRobj` constraint):
```Rust
enum AllowNull<T>
where T : FromRobj<T> {
    Null,
    NotNull(T)
}
```

With these things set up and agreed on, we can start contributing and polishing the type system.
